Crash when backend connection from another hostgroup is closed by MySQL while session is in FAST_FORWARD mode

# Bug Report

**ProxySQL Version:** version 3.0.6-921-g3803f11

**OS Version:** Ubuntu 13.3.0-6 ubuntu2~24.04.1 (x86_64)

**Backend Database:** MySQL 5.7

**Client Driver:** go-mysql-org/go-mysql

## Description

ProxySQL crashes with assertion failure when handling `COM_BINLOG_DUMP` command. The crash occurs in `MySQL_Connection::handler()` at `mysql_connection.cpp:1904` because `ASYNC_IDLE` state is not handled in the switch-case.


## Steps to Reproduce

1. Configure ProxySQL to connect to the MySQL 5.7 backend.

2. Use the `script` to send the `COM_BINLOG_DUMP` command to start binlog replication.

3. At this point, the proxy has one client connection and corresponding two backend MySQL connections responsible for reading and writing.

4. The backend connection responsible for reading is interrupted by MySQL due to `wait_timeout`.

5. The proxy experiences a core error while processing the connection via POLLIN.

**Reproduction Script (Go):**

```go
package main

import (
	"context"
	"flag"
	"fmt"
	"os"
	"os/signal"
	"syscall"

	"github.com/go-mysql-org/go-mysql/client"
	"github.com/go-mysql-org/go-mysql/mysql"
	"github.com/go-mysql-org/go-mysql/replication"
)

func main() {
	host := flag.String("host", "127.0.0.1", "MySQL host")
	port := flag.Int("port", 3306, "MySQL port")
	user := flag.String("user", "admin", "MySQL user")
	password := flag.String("password", "xxxxxx", "MySQL password")
	serverID := flag.Uint("server-id", 100, "Replication server ID")
	flag.Parse()

	// Get current binlog position
	conn, err := client.Connect(fmt.Sprintf("%s:%d", *host, *port), *user, *password, "sbtest")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Connection failed: %v\n", err)
		os.Exit(1)
	}

	result, err := conn.Execute("SHOW MASTER STATUS")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to get master status: %v\n", err)
		os.Exit(1)
	}

	var binlogFile string
	var binlogPos uint32
	if len(result.Values) > 0 {
		binlogFile = string(result.Values[0][0].AsString())
		binlogPos = uint32(result.Values[0][1].AsUint64())
	}
	conn.Close()

	fmt.Printf("Using Position mode (COM_BINLOG_DUMP)\n")
	fmt.Printf("Current binlog file: %s, position: %d\n", binlogFile, binlogPos)

	// Configure BinlogSyncer with Option callback to execute init SQL
	syncerCfg := replication.BinlogSyncerConfig{
		ServerID:        uint32(*serverID),
		Flavor:          "mysql",
		Host:            *host,
		Port:            uint16(*port),
		User:            *user,
		Password:        *password,
		HeartbeatPeriod: 30,
		// Option callback: execute init SQL on the same connection before sending COM_BINLOG_DUMP
		Option: func(c *client.Conn) error {
			// 1. Switch database
			if _, err := c.Execute("USE sbtest"); err != nil {
				return fmt.Errorf("USE sbtest failed: %w", err)
			}
			fmt.Println("[Option] Executed: USE sbtest")
			
			// 2. set SQL_LOG_BIN
			if _, err := c.Execute("select SQL_CALC_FOUND_ROWS * from lmt_0326"); err != nil {
				return fmt.Errorf("SELECT failed: %w", err)
			}
			fmt.Println("[Option] Executed: select SQL_CALC_FOUND_ROWS * from lmt_0326")

			// 2. Execute INSERT
			if _, err := c.Execute("INSERT INTO lmt_0326 VALUES (1);"); err != nil {
				return fmt.Errorf("INSERT into lmt_0326 failed: %w", err)
			}
			fmt.Println("[Option] Executed: INSERT INTO lmt_0326 VALUES (1);")

			// 3. Execute SELECT
			result, err := c.Execute("SELECT COUNT(1) FROM sbtest5")
			if err != nil {
				return fmt.Errorf("SELECT from sbtest5 failed: %w", err)
			}
			if len(result.Values) > 0 {
				count := int(result.Values[0][0].AsInt64())
				fmt.Printf("[Option] SELECT COUNT(1) FROM sbtest5: %d\n", count)
			}

			return nil
		},
	}
	syncer := replication.NewBinlogSyncer(syncerCfg)

	// StartSync establishes a new connection, calls Option callback to execute init SQL, then sends COM_BINLOG_DUMP
	streamer, err := syncer.StartSync(mysql.Position{
		Name: binlogFile,
		Pos:  binlogPos,
	})
	if err != nil {
		fmt.Fprintf(os.Stderr, "StartSync failed: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Connected, waiting for binlog events...")

	ctx, cancel := context.WithCancel(context.Background())
	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigCh
		fmt.Println("\nReceived exit signal, closing connection...")
		cancel()
	}()

	for {
		select {
		case <-ctx.Done():
			syncer.Close()
			fmt.Println("Program exited")
			return
		default:
			ev, err := streamer.GetEvent(ctx)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Failed to get event: %v\n", err)
				return
			}
			if ev.Header.EventType == replication.HEARTBEAT_EVENT {
				continue
			}
			fmt.Printf("Event type: %s, timestamp: %d, position: %d\n",
				ev.Header.EventType.String(),
				ev.Header.Timestamp,
				ev.Header.LogPos)
		}
	}
}

```

## Actual Behavior

ProxySQL crashes when handling the connection. The following crash logs were observed:

```
proxysql: mysql_connection.cpp:1985: MDB_ASYNC_ST MySQL_Connection::handler(short int): Assertion `0' failed.
Error: signal 6:
/data/proxysql306/proxysql/bin/proxysql(_Z13crash_handleri+0x41)[0x6337ee03aa01]
/lib/x86_64-linux-gnu/libc.so.6(+0x45330)[0x737fd9445330]
/lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x11c)[0x737fd949eb2c]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0x1e)[0x737fd944527e]
/lib/x86_64-linux-gnu/libc.so.6(abort+0xdf)[0x737fd94288ff]
/lib/x86_64-linux-gnu/libc.so.6(+0x2881b)[0x737fd942881b]
/lib/x86_64-linux-gnu/libc.so.6(+0x3b517)[0x737fd943b517]
/data/proxysql306/proxysql/bin/proxysql(_ZN16MySQL_Connection7handlerEs+0x20e)[0x6337ee435abe]
/data/proxysql306/proxysql/bin/proxysql(_ZN12MySQL_Thread27process_data_on_data_streamEP17MySQL_Data_Streamj+0x53c)[0x6337ee0ae66c]
/data/proxysql306/proxysql/bin/proxysql(_ZN11Base_Thread24ProcessAllMyDS_AfterPollI12MySQL_ThreadEEvv+0x84)[0x6337ee28c454]
/data/proxysql306/proxysql/bin/proxysql(_ZN12MySQL_Thread3runEv+0x34c)[0x6337ee0be59c]
/data/proxysql306/proxysql/bin/proxysql(_Z24mysql_worker_thread_funcPv+0x8c)[0x6337ee02277c]
/lib/x86_64-linux-gnu/libc.so.6(+0x9caa4)[0x737fd949caa4]
/lib/x86_64-linux-gnu/libc.so.6(+0x129c6c)[0x737fd9529c6c]
 ---- /data/proxysql306/proxysql/bin/proxysql(_Z13crash_handleri+0x41) [0x6337ee03aa01] : crash_handler(int)
 ---- /data/proxysql306/proxysql/bin/proxysql(_ZN16MySQL_Connection7handlerEs+0x20e) [0x6337ee435abe] : MySQL_Connection::handler(short)
 ---- /data/proxysql306/proxysql/bin/proxysql(_ZN12MySQL_Thread27process_data_on_data_streamEP17MySQL_Data_Streamj+0x53c) [0x6337ee0ae66c] : MySQL_Thread::process_data_on_data_stream(MySQL_Data_Stream*, unsigned int)
 ---- /data/proxysql306/proxysql/bin/proxysql(_ZN11Base_Thread24ProcessAllMyDS_AfterPollI12MySQL_ThreadEEvv+0x84) [0x6337ee28c454] : void Base_Thread::ProcessAllMyDS_AfterPoll<MySQL_Thread>()
 ---- /data/proxysql306/proxysql/bin/proxysql(_ZN12MySQL_Thread3runEv+0x34c) [0x6337ee0be59c] : MySQL_Thread::run()
 ---- /data/proxysql306/proxysql/bin/proxysql(_Z24mysql_worker_thread_funcPv+0x8c) [0x6337ee02277c] : mysql_worker_thread_func(void*)
To report a crashing bug visit: https://github.com/sysown/proxysql/issues
For support visit: https://proxysql.com/services/support/
2026-04-01 14:31:28 main.cpp:1922:ProxySQL_daemonize_phase3(): [ERROR] ProxySQL crashed. Restarting!
```


**GDB variable analysis:**
Regardless of the order, only print the key status values.
```
(gdb) print async_state_machine
$1 = ASYNC_IDLE
(gdb) p ((MySQL_Backend **)myds->sess->mybes->pdata)[1]->server_myds->DSS
$6 = STATE_READY
(gdb) p ((MySQL_Backend **)myds->sess->mybes->pdata)[0]->server_myds->DSS
$7 = STATE_MARIADB_GENERIC
(gdb) p ((MySQL_Backend **)myds->sess->mybes->pdata)[0]->hostgroup_id
$8 = 2
(gdb) p ((MySQL_Backend **)myds->sess->mybes->pdata)[1]->hostgroup_id
$9 = 1
(gdb) p this->myds->sess->status
$36 = FAST_FORWARD
(gdb) p myds->sess->mybes->len
$1 = 2
(gdb) p myds->revents
$14 = 1
```

The `ASYNC_IDLE` state is not handled in the switch-case statement in `MySQL_Connection::handler()`, causing the assertion to fail.

## Root Cause Analysis

The crash occurs because `async_state_machine` is `ASYNC_IDLE` when entering the switch statement in `MySQL_Connection::handler()` at `mysql_connection.cpp:1904`. The `ASYNC_IDLE` state is not matched in the switch-case, leading to assertion failure.

However, the root cause is that when receiving `MYSQL_COM_BINLOG_DUMP_GTID`, one scenario was overlooked: `there are still backend connections from other hostgroups in the current session`.

## Expected Behavior

ProxySQL should handle the `POLLIN` signal gracefully without crashing, or return an appropriate error message.

## ProxySQL Configuration

proxy config
```
datadir="/data/proxysql306/node_6033/data"

errorlog="/data/proxysql306/node_6033/logs/proxysql.log"


admin_variables=
{
    admin_credentials="xxxxxx"
    mysql_ifaces="0.0.0.0:6032"

    restapi_enabled=true
    restapi_port=6070
    prometheus_memory_metrics_interval=30
}

mysql_variables=
{
        # Proxysql employs an IO multiplexing model, with multiple worker threads functioning as reactor threads.
        # Given that the CPU limit imposed on LXC containers is not a hard limit,
        # and that IO processing may incur waiting,
        # adding two based on current CPU cores can facilitate greater utilization
        threads=6
        max_connections=6400
        default_query_delay=0
        have_compress=true
        poll_timeout=2000
        interfaces="0.0.0.0:6033"
        default_schema="information_schema"
        stacksize=1048576
        server_version="8.0"

        monitor_username="monitor"
        monitor_password="monitor"
        monitor_history=600000
        monitor_connect_interval=60000
        monitor_ping_interval=10000
        monitor_read_only_interval=1500

        ping_interval_server_msec=120000
        ping_timeout_server=500
        commands_stats=true
        sessions_sort=true
        connect_retries_on_failure=10

        verbose_query_error=true

        monitor_user_sync_enabled=true
        monitor_user_sync_interval=5000
        monitor_user_sync_timeout=500

        default_max_latency_ms=2000
        shun_recovery_time_sec=3
        monitor_connect_timeout=3000
        monitor_ping_timeout=2000
        monitor_read_only_timeout=2000

        # Set the maximum number of prepared statements that each ProxySQL connection can cache.
        # Exceeding this limit may cause frequent resets of user sessions.
        max_stmts_per_connection=100
        # To address the transaction consistency issue in ProxySQL.
        # When set to false by default, the system overrides the user's autocommit=0 setting and forces it to 1.
        enforce_autocommit_on_reads=true

        monitor_slave_lag_when_null=604800
        monitor_replication_lag_count=3
        monitor_replication_lag_interval=2000
        request_fallback_to_default_group=true

        multiplexing=true
        connection_delay_multiplex_ms=0
        connect_timeout_client=15000
        connect_timeout_server_max=15000
        connect_timeout_server=5000

        max_allowed_packet=536870912
        default_query_timeout=1728000000
        max_transaction_time=1728000000
        max_transaction_idle_time=1728000000
}

mysql_servers =
(

    { address="xxxxxx" , port=3306 , hostgroup=1, max_connections=2048 },
    { address="xxxxxx" , port=3306 , hostgroup=2, max_connections=2048 },
)

# defines all the MySQL users
mysql_users:
(
    {
        username = "xxxx"
        password = "xxxx"
        default_hostgroup = 1
        active = 1
    }
)

mysql_query_rules:
(
    {
        rule_id=1
        active=1
        match_pattern="\/\*[[:space:]]*use_master[[:space:]]*\*\/"
        destination_hostgroup=1
        apply=1
    },
    {
        rule_id=2
        active=1
        match_digest="^SELECT.*FOR UPDATE"
        destination_hostgroup=1
        apply=1
    },
    {
        rule_id=3
        active=1
        match_digest="^SELECT.*LOCK IN SHARE MODE"
        destination_hostgroup=1
        apply=1
    },
    {
        rule_id=4
        active=1
        match_digest="^SELECT"
        destination_hostgroup=2
        apply=1
    }
)

```

--- a/lib/MySQL_Session.cpp
+++ b/lib/MySQL_Session.cpp
@@ -5101,6 +5101,33 @@ int MySQL_Session::GPFC_Replication_SwitchToFastForward(PtrSize_t& pkt, unsigned
 		set_status(FAST_FORWARD); // we can set status to FAST_FORWARD
 	}
 
+	// Issue #5556: Clean up idle backends from other hostgroups.
+	// When transitioning to FAST_FORWARD, only the active backend (mybe) is
+	// needed. Any other backends from different hostgroups are now idle and
+	// will never be used again. If left alive, they can receive POLLIN events
+	// (e.g. from MySQL wait_timeout) which causes a crash in handler() because
+	// ASYNC_IDLE is not handled in the switch-case.
+	{
+		unsigned int i = 0;
+		while (i < mybes->len) {
+			MySQL_Backend* _mybe = (MySQL_Backend*)mybes->index(i);
+			if (_mybe != mybe) {
+				if (_mybe->server_myds && _mybe->server_myds->myconn) {
+					proxy_info(
+						"Cleaning up idle backend from hostgroup %d during FAST_FORWARD transition\n",
+						_mybe->hostgroup_id
+					);
+					_mybe->server_myds->destroy_MySQL_Connection_From_Pool(false);
+				}
+				mybes->remove_index_fast(i);
+				_mybe->reset();
+				delete _mybe;
+			} else {
+				i++;
+			}
+		}
+	}
+
 	return 0;
 }
 

--- a/test/tap/groups/groups.json
+++ b/test/tap/groups/groups.json
@@ -252,6 +252,7 @@
   "test_auth_methods-t" : [ "mysql84-g2","mysql-auto_increment_delay_multiplex=0-g2","mysql-multiplexing=false-g2","mysql-query_digests=0-g2","mysql-query_digests_keep_comment=1-g2" ],
   "test_auto_increment_delay_multiplex-t" : [ "legacy-g2","mysql84-g2","mysql-auto_increment_delay_multiplex=0-g2","mysql-multiplexing=false-g2","mysql-query_digests=0-g2","mysql-query_digests_keep_comment=1-g2" ],
   "test_backend_conn_ping-t" : [ "legacy-g2","mysql84-g2","mysql-auto_increment_delay_multiplex=0-g2","mysql-multiplexing=false-g2","mysql-query_digests=0-g2","mysql-query_digests_keep_comment=1-g2" ],
+  "test_binlog_dump_multi_backend_crash-t" : [ "legacy-binlog-g1" ],
   "test_binlog_fast_forward-t" : [ "legacy-binlog-g1" ],
   "test_binlog_reader-t" : [ "legacy-binlog-g1" ],
   "test_binlog_reader_uses_previous_hostgroup-t" : [ "legacy-binlog-g1" ],

--- a/test/tap/tests/test_binlog_dump_multi_backend_crash-t.cpp
+++ b/test/tap/tests/test_binlog_dump_multi_backend_crash-t.cpp
@@ -91,28 +91,8 @@ int main(int argc, char** argv) {
 		return exit_status();
 	}
 
-	// Save and lower wait_timeout
-	long orig_wait_timeout = 28800;
-	if (mysql_query(backend, "SELECT @@global.wait_timeout") == 0) {
-		MYSQL_RES* res = mysql_store_result(backend);
-		if (res) {
-			MYSQL_ROW row = mysql_fetch_row(res);
-			if (row && row[0]) orig_wait_timeout = strtol(row[0], NULL, 10);
-			mysql_free_result(res);
-		}
-	}
-	diag("Original wait_timeout: %ld, setting to %d", orig_wait_timeout, SHORT_WAIT_TIMEOUT);
-
-	std::string set_wt = "SET GLOBAL wait_timeout=" + std::to_string(SHORT_WAIT_TIMEOUT);
-	if (mysql_query(backend, set_wt.c_str()) != 0) {
-		diag("Failed to set wait_timeout: %s", mysql_error(backend));
-		mysql_close(backend);
-		mysql_close(admin);
-		return exit_status();
-	}
-
 	// ========================================================================
-	// Step 2: Discover hostgroup topology and set up routing
+	// Step 2: Discover hostgroup topology
 	// ========================================================================
 	int writer_hg = -1;
 	int reader_hg = -1;
@@ -139,6 +119,27 @@ int main(int argc, char** argv) {
 		return exit_status();
 	}
 	diag("Discovered hostgroups: writer=%d, reader=%d", writer_hg, reader_hg);
+
+	// Save and lower wait_timeout (after skip checks so early returns don't
+	// leave it modified — addresses Copilot review feedback)
+	long orig_wait_timeout = 28800;
+	if (mysql_query(backend, "SELECT @@global.wait_timeout") == 0) {
+		MYSQL_RES* res = mysql_store_result(backend);
+		if (res) {
+			MYSQL_ROW row = mysql_fetch_row(res);
+			if (row && row[0]) orig_wait_timeout = strtol(row[0], NULL, 10);
+			mysql_free_result(res);
+		}
+	}
+	diag("Original wait_timeout: %ld, setting to %d", orig_wait_timeout, SHORT_WAIT_TIMEOUT);
+
+	std::string set_wt = "SET GLOBAL wait_timeout=" + std::to_string(SHORT_WAIT_TIMEOUT);
+	if (mysql_query(backend, set_wt.c_str()) != 0) {
+		diag("Failed to set wait_timeout: %s", mysql_error(backend));
+		mysql_close(backend);
+		mysql_close(admin);
+		return exit_status();
+	}
 
 	// Set up query rules: ^SELECT → reader_hg (multiplex=0), all else → writer_hg (default)
 	MYSQL_QUERY(admin, "DELETE FROM mysql_query_rules");
@@ -191,6 +192,37 @@ int main(int argc, char** argv) {
 	if (mysql_query(proxy, "DO 1") != 0) {
 		diag("DO 1 failed: %s", mysql_error(proxy));
 		goto cleanup;
+	}
+
+	// Verify connections were established to both hostgroups.
+	// We check ConnOK (total connections ever made) rather than ConnUsed,
+	// because the writer backend may already be returned to the pool after
+	// DO completes (it's still in the session's mybes array though).
+	{
+		bool has_writer = false, has_reader = false;
+		std::string pool_q = "SELECT hostgroup, ConnOK FROM stats_mysql_connection_pool "
+			"WHERE hostgroup IN (" + std::to_string(writer_hg) + "," + std::to_string(reader_hg) + ") "
+			"AND ConnOK > 0";
+		if (mysql_query(admin, pool_q.c_str()) == 0) {
+			MYSQL_RES* res = mysql_store_result(admin);
+			if (res) {
+				MYSQL_ROW row;
+				while ((row = mysql_fetch_row(res))) {
+					int hg = atoi(row[0]);
+					diag("  Connection pool: HG%d has %s total connections (ConnOK)", hg, row[1]);
+					if (hg == writer_hg) has_writer = true;
+					if (hg == reader_hg) has_reader = true;
+				}
+				mysql_free_result(res);
+			}
+		}
+		if (!has_writer || !has_reader) {
+			diag("PRECONDITION FAILED: Connections not established to both hostgroups "
+				"(writer=%s, reader=%s). Test cannot reproduce #5556.",
+				has_writer ? "yes" : "NO", has_reader ? "yes" : "NO");
+			goto cleanup;
+		}
+		diag("Confirmed connections to both writer HG%d and reader HG%d", writer_hg, reader_hg);
 	}
 
 	// ========================================================================

--- a/test/tap/tests/test_binlog_dump_multi_backend_crash-t.cpp
+++ b/test/tap/tests/test_binlog_dump_multi_backend_crash-t.cpp
@@ -1,0 +1,339 @@
+/**
+ * @file test_binlog_dump_multi_backend_crash-t.cpp
+ * @brief Regression test for issue #5556.
+ * @details When a session has backend connections to multiple hostgroups and
+ *   then receives COM_BINLOG_DUMP, ProxySQL switches the session to
+ *   FAST_FORWARD mode but only manages the active backend. The idle backend
+ *   from another hostgroup remains in the session. If that idle backend
+ *   receives a POLLIN event (e.g. from MySQL wait_timeout closing the
+ *   connection), MySQL_Connection::handler() is called with
+ *   async_state_machine==ASYNC_IDLE, which has no case in the switch and
+ *   hits assert(0).
+ *
+ *   Reproduction steps:
+ *   1. Use the existing replication hostgroups (writer_hg / reader_hg).
+ *      Add a query rule routing SELECTs to reader_hg with multiplex=0.
+ *   2. Disable set_query_lock_on_hostgroup so SET @user_var doesn't
+ *      prevent cross-hostgroup routing.
+ *   3. Connect to ProxySQL and run SELECT (creates reader_hg backend)
+ *      then DO 1 (creates writer_hg backend).
+ *   4. Send COM_BINLOG_DUMP via mariadb_rpl (uses writer_hg via
+ *      previous_hostgroup).
+ *   5. Wait for MySQL wait_timeout to close the idle reader_hg backend.
+ *   6. Verify ProxySQL did not crash.
+ */
+
+#include <cstdlib>
+#include <cstdio>
+#include <cstring>
+#include <unistd.h>
+#include <string>
+#include <vector>
+
+#include "mysql.h"
+#include "mariadb_rpl.h"
+
+#include "tap.h"
+#include "command_line.h"
+#include "utils.h"
+
+// Short wait_timeout to trigger idle backend closure quickly
+#define SHORT_WAIT_TIMEOUT 5
+// How long to sleep waiting for the idle backend to be closed by MySQL
+#define SLEEP_SECONDS (SHORT_WAIT_TIMEOUT + 3)
+
+int main(int argc, char** argv) {
+	CommandLine cl;
+
+	plan(1);
+	diag("Testing issue #5556: COM_BINLOG_DUMP crash with idle backends from other hostgroups");
+
+	if (cl.getEnv()) {
+		diag("Failed to get the required environmental variables.");
+		return exit_status();
+	}
+
+	// ========================================================================
+	// Step 0: Connect to ProxySQL admin
+	// ========================================================================
+	MYSQL* admin = mysql_init(NULL);
+	if (!admin || !mysql_real_connect(admin, cl.admin_host, cl.admin_username,
+			cl.admin_password, NULL, cl.admin_port, NULL, 0)) {
+		fprintf(stderr, "Admin connection failed: %s\n", admin ? mysql_error(admin) : "mysql_init failed");
+		return exit_status();
+	}
+
+	// ========================================================================
+	// Step 1: Connect to MySQL backend directly, check binlog, set wait_timeout
+	// ========================================================================
+	MYSQL* backend = mysql_init(NULL);
+	if (!backend || !mysql_real_connect(backend, cl.mysql_host, cl.mysql_username,
+			cl.mysql_password, NULL, cl.mysql_port, NULL, 0)) {
+		fprintf(stderr, "Backend connection failed: %s\n", backend ? mysql_error(backend) : "mysql_init failed");
+		mysql_close(admin);
+		return exit_status();
+	}
+
+	// Verify binlog is available
+	bool binlog_ok = false;
+	if (mysql_query(backend, "SHOW MASTER STATUS") == 0) {
+		MYSQL_RES* res = mysql_store_result(backend);
+		if (res) {
+			if (mysql_fetch_row(res)) binlog_ok = true;
+			mysql_free_result(res);
+		}
+	}
+	if (!binlog_ok) {
+		diag("Binary logging not available. Skipping.");
+		ok(1, "SKIP: Binary logging not available");
+		mysql_close(backend);
+		mysql_close(admin);
+		return exit_status();
+	}
+
+	// Save and lower wait_timeout
+	long orig_wait_timeout = 28800;
+	if (mysql_query(backend, "SELECT @@global.wait_timeout") == 0) {
+		MYSQL_RES* res = mysql_store_result(backend);
+		if (res) {
+			MYSQL_ROW row = mysql_fetch_row(res);
+			if (row && row[0]) orig_wait_timeout = strtol(row[0], NULL, 10);
+			mysql_free_result(res);
+		}
+	}
+	diag("Original wait_timeout: %ld, setting to %d", orig_wait_timeout, SHORT_WAIT_TIMEOUT);
+
+	std::string set_wt = "SET GLOBAL wait_timeout=" + std::to_string(SHORT_WAIT_TIMEOUT);
+	if (mysql_query(backend, set_wt.c_str()) != 0) {
+		diag("Failed to set wait_timeout: %s", mysql_error(backend));
+		mysql_close(backend);
+		mysql_close(admin);
+		return exit_status();
+	}
+
+	// ========================================================================
+	// Step 2: Discover hostgroup topology and set up routing
+	// ========================================================================
+	int writer_hg = -1;
+	int reader_hg = -1;
+	{
+		const char* q = "SELECT writer_hostgroup, reader_hostgroup "
+			"FROM mysql_replication_hostgroups LIMIT 1";
+		if (mysql_query(admin, q) == 0) {
+			MYSQL_RES* res = mysql_store_result(admin);
+			if (res) {
+				MYSQL_ROW row = mysql_fetch_row(res);
+				if (row && row[0] && row[1]) {
+					writer_hg = atoi(row[0]);
+					reader_hg = atoi(row[1]);
+				}
+				mysql_free_result(res);
+			}
+		}
+	}
+	if (writer_hg < 0 || reader_hg < 0) {
+		diag("Could not discover replication hostgroups. Skipping.");
+		ok(1, "SKIP: No replication hostgroups configured");
+		mysql_close(backend);
+		mysql_close(admin);
+		return exit_status();
+	}
+	diag("Discovered hostgroups: writer=%d, reader=%d", writer_hg, reader_hg);
+
+	// Set up query rules: ^SELECT → reader_hg (multiplex=0), all else → writer_hg (default)
+	MYSQL_QUERY(admin, "DELETE FROM mysql_query_rules");
+	{
+		std::string q = "INSERT INTO mysql_query_rules "
+			"(rule_id, active, match_digest, destination_hostgroup, multiplex, apply) "
+			"VALUES (1, 1, '^SELECT', " + std::to_string(reader_hg) + ", 0, 1)";
+		MYSQL_QUERY(admin, q.c_str());
+	}
+	MYSQL_QUERY(admin, "LOAD MYSQL QUERY RULES TO RUNTIME");
+
+	// Disable hostgroup locking so SET @user_var queries don't prevent
+	// routing to different hostgroups within the same session
+	MYSQL_QUERY(admin, "SET mysql-set_query_lock_on_hostgroup=0");
+	MYSQL_QUERY(admin, "LOAD MYSQL VARIABLES TO RUNTIME");
+
+	diag("ProxySQL configured: SELECT→HG%d (reader), default→HG%d (writer), multiplex=0, lock_on_hg=0",
+		reader_hg, writer_hg);
+
+	// ========================================================================
+	// Step 3: Connect to ProxySQL and create backends in both hostgroups
+	// ========================================================================
+	MYSQL* proxy = mysql_init(NULL);
+	if (!proxy) {
+		diag("mysql_init failed for proxy connection");
+		goto cleanup;
+	}
+	// Force CLIENT_DEPRECATE_EOF to match backend capabilities for fast-forward
+	proxy->options.client_flag |= CLIENT_DEPRECATE_EOF;
+
+	if (!mysql_real_connect(proxy, cl.root_host, cl.root_username,
+			cl.root_password, NULL, cl.root_port, NULL, 0)) {
+		diag("Proxy connection failed: %s", mysql_error(proxy));
+		goto cleanup;
+	}
+
+	// SELECT → creates backend in reader_hg
+	diag("Sending SELECT to create backend in reader HG%d", reader_hg);
+	if (mysql_query(proxy, "SELECT 1") != 0) {
+		diag("SELECT failed: %s", mysql_error(proxy));
+		goto cleanup;
+	}
+	{
+		MYSQL_RES* res = mysql_store_result(proxy);
+		mysql_free_result(res);
+	}
+
+	// DO 1 → creates backend in writer_hg, also sets previous_hostgroup
+	diag("Sending DO to create backend in writer HG%d (and set previous_hostgroup)", writer_hg);
+	if (mysql_query(proxy, "DO 1") != 0) {
+		diag("DO 1 failed: %s", mysql_error(proxy));
+		goto cleanup;
+	}
+
+	// ========================================================================
+	// Step 4: Send COM_BINLOG_DUMP via mariadb_rpl
+	// ========================================================================
+	{
+		// Set up replication parameters on the same connection
+		diag("Setting replication parameters");
+		if (mysql_query(proxy, "SET @master_binlog_checksum='NONE'") != 0) {
+			diag("SET checksum failed: %s", mysql_error(proxy));
+			goto cleanup;
+		}
+		if (mysql_query(proxy, "SET @master_heartbeat_period=1000000000") != 0) {
+			diag("SET heartbeat failed: %s", mysql_error(proxy));
+			goto cleanup;
+		}
+
+		MARIADB_RPL* rpl = mariadb_rpl_init(proxy);
+		if (!rpl) {
+			diag("mariadb_rpl_init failed");
+			goto cleanup;
+		}
+		rpl->server_id = 99999; // unique server ID for test
+		rpl->start_position = 4; // start from beginning of first binlog
+		rpl->flags = MARIADB_RPL_BINLOG_SEND_ANNOTATE_ROWS;
+
+		diag("Sending COM_BINLOG_DUMP: pos=4 server_id=%d", rpl->server_id);
+		if (mariadb_rpl_open(rpl)) {
+			diag("mariadb_rpl_open failed: %s", mysql_error(proxy));
+			mariadb_rpl_close(rpl);
+			goto cleanup;
+		}
+
+		// Fetch a couple of events to confirm replication is working
+		diag("Replication started, fetching initial events...");
+		MARIADB_RPL_EVENT* event = NULL;
+		int num_events = 0;
+		for (int i = 0; i < 5; i++) {
+			event = mariadb_rpl_fetch(rpl, event);
+			if (!event) break;
+			num_events++;
+			diag("  Event %d: type=%d", num_events, event->event_type);
+		}
+		if (event) mariadb_free_rpl_event(event);
+		diag("Received %d initial events", num_events);
+
+		// ====================================================================
+		// Step 5: Wait for the idle backend (reader_hg) to be closed by MySQL
+		// ====================================================================
+		diag("Sleeping %d seconds for wait_timeout to close idle reader HG%d backend...",
+			SLEEP_SECONDS, reader_hg);
+		sleep(SLEEP_SECONDS);
+
+		// Fetch one more event to force ProxySQL to process the POLLIN on the
+		// idle backend's fd. This is the point where the crash would occur.
+		diag("Fetching another event (this triggers processing of idle backend POLLIN)...");
+		event = mariadb_rpl_fetch(rpl, NULL);
+		if (event) {
+			diag("  Got event: type=%d", event->event_type);
+			mariadb_free_rpl_event(event);
+		}
+
+		mariadb_rpl_close(rpl);
+	}
+
+	// ========================================================================
+	// Step 6: Verify ProxySQL did not crash
+	// ========================================================================
+	{
+		// Try to query the admin interface. If ProxySQL crashed and restarted,
+		// our admin connection would be broken.
+		int admin_ok = 0;
+
+		// First check our existing admin connection
+		if (mysql_query(admin, "SELECT 1") == 0) {
+			MYSQL_RES* res = mysql_store_result(admin);
+			if (res) {
+				mysql_free_result(res);
+				admin_ok = 1;
+			}
+		}
+
+		// If the existing connection is broken, try a fresh one
+		if (!admin_ok) {
+			diag("Existing admin connection broken, trying fresh connection...");
+			MYSQL* admin2 = mysql_init(NULL);
+			if (admin2 && mysql_real_connect(admin2, cl.admin_host, cl.admin_username,
+					cl.admin_password, NULL, cl.admin_port, NULL, 0)) {
+				// ProxySQL restarted after crash - the bug is present
+				diag("ProxySQL appears to have restarted (crash detected)");
+				mysql_close(admin2);
+				admin_ok = 0;
+			} else {
+				diag("Cannot connect to ProxySQL admin at all");
+				admin_ok = 0;
+			}
+		}
+
+		ok(admin_ok, "ProxySQL should not crash when idle backend receives POLLIN during FAST_FORWARD (issue #5556)");
+	}
+
+cleanup:
+	// Restore original wait_timeout
+	{
+		std::string restore_wt = "SET GLOBAL wait_timeout=" + std::to_string(orig_wait_timeout);
+		if (mysql_query(backend, restore_wt.c_str()) != 0) {
+			mysql_close(backend);
+			backend = mysql_init(NULL);
+			if (backend && mysql_real_connect(backend, cl.mysql_host, cl.mysql_username,
+					cl.mysql_password, NULL, cl.mysql_port, NULL, 0)) {
+				mysql_query(backend, restore_wt.c_str());
+			}
+		}
+		diag("Restored wait_timeout to %ld", orig_wait_timeout);
+	}
+
+	// Clean up query rules and ProxySQL variables
+	{
+		MYSQL* admin_cleanup = admin;
+		bool need_new = false;
+		if (mysql_query(admin, "SELECT 1") != 0) {
+			need_new = true;
+			admin_cleanup = mysql_init(NULL);
+			if (admin_cleanup) {
+				mysql_real_connect(admin_cleanup, cl.admin_host, cl.admin_username,
+					cl.admin_password, NULL, cl.admin_port, NULL, 0);
+			}
+		} else {
+			MYSQL_RES* res = mysql_store_result(admin);
+			mysql_free_result(res);
+		}
+
+		if (admin_cleanup) {
+			MYSQL_QUERY_err(admin_cleanup, "SET mysql-set_query_lock_on_hostgroup=1");
+			MYSQL_QUERY_err(admin_cleanup, "LOAD MYSQL VARIABLES TO RUNTIME");
+			MYSQL_QUERY_err(admin_cleanup, "DELETE FROM mysql_query_rules");
+			MYSQL_QUERY_err(admin_cleanup, "LOAD MYSQL QUERY RULES TO RUNTIME");
+			if (need_new) mysql_close(admin_cleanup);
+		}
+	}
+
+	mysql_close(backend);
+	mysql_close(admin);
+
+	return exit_status();
+}

